### PR TITLE
feat: sync Shopify payouts [CU-1k99w5]

### PR DESCRIPTION
**Pull request by Alchez**
**Changes:**

- (Feature) Added a new **Shopify Payout** doctype:
	- Summary of individual Shopify Payouts.
	- List of transactions in individual Payouts.
	- On submit, creates a Journal Entry to balance the invoices.
- (Feature) Added a daily hook to **sync new Payouts** from the connected Shopify instance.
- (Feature) Added new **Sync** buttons in Shopify Settings to sync Shopify items and payouts.
- (Enhancement) Use [Shopify's python library](https://github.com/Shopify/shopify_python_api) instead of making requests.
- (Enhancement) Cleanup the existing integration code.

<hr>

**Screenshots / GIFs:**

image

### CODE REVIEW

**Comment by vjFaLk** [comment]
Task linked: [CU-dyv5zx Investigate Shopify Payment API](https://app.clickup.com/t/dyv5zx)

**Comment by vjFaLk** [comment]
Task linked: [CU-1k7v8t Poll Shopify's API to retrieve payouts and make payments every day](https://app.clickup.com/t/1k7v8t)

**Comment by vjFaLk** [comment]
Task linked: [CU-1k7uu2 Add disclaimer in Shopify Settings to enable "Shopify Payments" on the Shopify admin and private app pages](https://app.clickup.com/t/1k7uu2)

File: erpnext/erpnext_integrations/connectors/shopify_connection.py [hunk 1]
```diff
@@ -18,90 +22,148 @@ def store_request_data(order=None, event=None):
...
+	Args:
+
+		order (dict): The Shopify order data.
+		request_id (str, optional): The ID of the existing Shopify Log document
+			for this request. Defaults to None.
+	"""
+
 	frappe.set_user('Administrator')
-	shopify_settings = frappe.get_doc("Shopify Settings")
+	frappe.flags.request_id = request_id
```
Comment (vjFaLk) [inline erpnext/erpnext_integrations/connectors/shopify_connection.py:42]: What's this about?

File: erpnext/erpnext_integrations/connectors/shopify_connection.py [hunk 2]
```diff
@@ -18,90 +22,148 @@ def store_request_data(order=None, event=None):
...
+		make_shopify_log(status="Error", response_data=order, exception=e)
+	else:
+		make_shopify_log(status="Success", response_data=order)
+		return so
+
+
+def create_shopify_invoice(order, so, request_id=None):
+	frappe.flags.request_id = request_id
+
+	if not order.get("financial_status") in ["paid", "partially_refunded", "refunded"]:
```
Comment (vjFaLk) [inline erpnext/erpnext_integrations/connectors/shopify_connection.py:75]: Should we do these check in `sync_sales_order` instead? In fact, the function name is also inaccurate as it's not just the Order is it? Maybe it should be `sync_shopify_data`?

File: erpnext/erpnext_integrations/connectors/shopify_connection.py [hunk 3]
```diff
@@ -18,90 +22,148 @@ def store_request_data(order=None, event=None):
...
-			make_shopify_log(status="Success")
 
 def prepare_sales_invoice(order, request_id=None):
 	frappe.set_user('Administrator')
-	shopify_settings = frappe.get_doc("Shopify Settings")
 	frappe.flags.request_id = request_id
 
 	try:
-		sales_order = get_sales_order(cstr(order['id']))
+		sales_order = get_shopify_document("Sales Order", cstr(order['id']))
```
Comment (vjFaLk) [inline erpnext/erpnext_integrations/connectors/shopify_connection.py]: `order.get()` for safety

File: erpnext/erpnext_integrations/connectors/shopify_connection.py [hunk 4]
```diff
@@ -132,98 +194,159 @@ def create_sales_order(shopify_order, shopify_settings, company=None):
...
+				shopify_product_id = frappe.db.get_value("Item", item.item_code, "shopify_product_id")
+				shopify_variant_id = frappe.db.get_value("Item", item.item_code, "shopify_variant_id")
+				if shopify_product_id in refunded_items or shopify_variant_id in refunded_variants:
+					continue
+
+				# set item values for non-refunded items to zero;
+				# preferring this over removal of the item to avoid zero-item
+				# refunds and downstream effects for other documents
+				item.qty = 0
+				item.discount_percentage = 100
```
Comment (vjFaLk) [inline erpnext/erpnext_integrations/connectors/shopify_connection.py:278]: Why discount, why not just rate?

**Review by vjFaLk — CHANGES_REQUESTED** [review]
Partial review.

**Comment by vjFaLk** [comment]
Task linked: [CU-1k99w5 Poll Shopify's API to retrieve payouts and make payments every day](https://app.clickup.com/t/1k99w5)

File: erpnext/erpnext_integrations/connectors/shopify_connection.py (hunk 1 above)
Comment (Alchez) [reply erpnext/erpnext_integrations/connectors/shopify_connection.py:42]: If a request fails, they use scheduler events to retry. The request ID is used as context here.

File: erpnext/erpnext_integrations/connectors/shopify_connection.py (hunk 4 above)
Comment (Alchez) [reply erpnext/erpnext_integrations/connectors/shopify_connection.py:278]: It gets overriden on validate for some reason. Discounts are the only way to maintain the zero rate.

File: erpnext/erpnext_integrations/connectors/shopify_connection.py [hunk 5]
```diff
@@ -132,98 +194,159 @@ def create_sales_order(shopify_order, shopify_settings, company=None):
...
+		shopify_order_id (int): The Shopify order ID.
+		shopify_financial_status (str): The financial status of the Shopify order.
+			Should be one of: refunded, partially_refunded.
+		sales_invoice (SalesInvoice): The Sales Invoice document.
+
+	Returns:
+		SalesInvoice: The Sales Invoice return document.
+	"""
+
+	shopify_settings = frappe.get_doc("Shopify Settings")
```
Comment (vjFaLk) [inline erpnext/erpnext_integrations/connectors/shopify_connection.py]: Use `get_single`

**Comment by Alchez** [comment]
Moved to https://github.com/ParsimonyGit/shopify_integration.

**Pull request closed without merging**